File: solutions/csharp/sieve/2/Sieve.cs
using System;
using System.Collections.Generic;

public static class Sieve
{
    public static int[] Primes(int limit)
    {
        if (limit < 1)
            throw new ArgumentOutOfRangeException("Argument must be a positive integer.");
        
        var sieve = new bool[limit + 1];
        var primes = new List<int>();

        for (int i = 2; i <= limit; i++)
        {
            if (sieve[i])
                continue;

            primes.Add(i);
            
            for (int j = 2 * i; j <= limit; j += i)
                sieve[j] = true;
        }

        return primes.ToArray();
    }
}

File: solutions/fortran/sieve/2/sieve.f90
module sieve
  implicit none

contains

  function primes(limit) result(array)
    integer, intent(in) :: limit
    integer, allocatable :: array(:)
    logical, allocatable :: sieve(:)
    integer :: i, j, count

    allocate(sieve(limit))
    sieve = .true.
    
    sieve(1) = .false.

    do i = 2, int(sqrt(real(limit)))
      if (sieve(i)) then
        do j = i*i, limit, i
          sieve(j) = .false.
        end do
      end if
    end do

    count = 0
    do i = 2, limit
      if (sieve(i)) count = count + 1
    end do

    allocate(array(count))
    count = 0

    do i = 2, limit
      if (sieve(i)) then
        count = count + 1
        array(count) = i
      end if
    end do

  end function primes

end module sieve
